src/node-capnp/capnp.cc

```diff
@@ -2521,6 +2521,173 @@ void throw_(const v8::FunctionCallbackInfo<v8::Value>& args) {
+    } else {
+      // Yay, array is aligned.
+      words = kj::arrayPtr(reinterpret_cast<const capnp::word*>(bytes.begin()),
+                           bytes.size() / sizeof(capnp::word));
+    }
+  }
+
+  inline const kj::ArrayPtr<const capnp::word>* operator->() const { return &words; }
+  inline const kj::ArrayPtr<const capnp::word>& operator*() const { return words; }
+
+private:
+  kj::ArrayPtr<const capnp::word> words;
+  kj::Array<capnp::word> copy;
+};
+
+bool matchPowerboxQuery(capnp::AnyPointer::Reader query, capnp::AnyPointer::Reader candidate);
+
+bool matchPowerboxQuery(capnp::AnyStruct::Reader query, capnp::AnyStruct::Reader candidate) {
+  {
+    // Compare data.
+    auto queryData = query.getDataSection();
+    auto candidateData = candidate.getDataSection();
+
+    auto commonSize = kj::min(queryData.size(), candidateData.size());
+    if (memcmp(queryData.begin(), candidateData.begin(), commonSize) != 0) {
+      // Data sections don't match.
+      return false;
+    }
+
+    // Non-matched parts of data sections must be all-zero.
+    kj::byte accum = 0;
+    for (kj::byte b: queryData.slice(commonSize, queryData.size())) {
+      accum |= b;
+    }
+    for (kj::byte b: candidateData.slice(commonSize, candidateData.size())) {
+      accum |= b;
+    }
+    if (accum != 0) return false;
+  }
+
+  {
+    // Compare pointers.
+    auto queryPointers = query.getPointerSection();
+    auto candidatePointers = candidate.getPointerSection();
+
+    auto commonSize = kj::min(queryPointers.size(), candidatePointers.size());
+    for (auto i: kj::range<decltype(commonSize)>(0, commonSize)) {
+      if (!matchPowerboxQuery(queryPointers[i], candidatePointers[i])) {
+        return false;
+      }
+    }
+
+    // No need to compare the non-overlapping range since null pointers match anything.
+  }
+
+  return true;
+}
+
+bool matchPowerboxQuery(capnp::AnyList::Reader query, capnp::AnyList::Reader candidate) {
+  auto elementSize = query.getElementSize();
+  if (candidate.getElementSize() != elementSize) return false;
+
+  switch (elementSize) {
+    case capnp::ElementSize::VOID:
+    case capnp::ElementSize::BIT:
+    case capnp::ElementSize::BYTE:
+    case capnp::ElementSize::TWO_BYTES:
+    case capnp::ElementSize::FOUR_BYTES:
+    case capnp::ElementSize::EIGHT_BYTES:
+      return query == candidate;
+
+    case capnp::ElementSize::POINTER: {
+      auto queryPtrs = query.as<capnp::List<capnp::AnyPointer>>();
+      auto candidatePtrs = candidate.as<capnp::List<capnp::AnyPointer>>();
+
+      if (queryPtrs.size() != candidatePtrs.size()) return false;
+      for (auto i: kj::indices(queryPtrs)) {
+        if (!matchPowerboxQuery(queryPtrs[i], candidatePtrs[i])) return false;
+      }
+      return true;
+    }
+
+    case capnp::ElementSize::INLINE_COMPOSITE: {
+      // Every element in the query must be matched by at least one element in the candidate.
+      //
+      // TODO(perf): O(m*n), can we do better?
```

Comment:
Ah, I suppose that `ReaderOptions.traversalLimitInWords` will cause an exception to be thrown at some point.